util/settings.lua

```diff
@@ -45,14 +45,63 @@ mail.settings = {
         type = "string", default = "%Y-%m-%d %X", group = "other", index = 3, label = S("Date format"),
         dataset = {"%Y-%m-%d %X", "%d/%m/%y %X", "%A %d %B %Y %X"}, format = os.date
     },
+    mute_list = {
+        type = "list", default = {}, group = "spam", index = 1,
+        label = S("Mute list")
+    },
 }
 
 mail.settings_groups = {
     { name = "notifications", label = S("Notifications")},
     { name = "message_list",  label = S("Message list")},
+    { name = "spam",          label = S("Spam")},
     { name = "other",         label = S("Other")}
 }
 
-for s, _ in pairs(mail.settings) do
+for s, d in pairs(mail.settings) do
 	mail.selected_idxs[s] = {}
+	if d.type == "list" then
+        mail.selected_idxs["index_" .. s] = {}
+    end
+end
+
+function mail.settings.mute_list.check(name, value)
+    local valid_players = {}
+    for _, p in ipairs(value) do
+        if p ~= name and minetest.player_exists(p) then
+            table.insert(valid_players, p)
+        end
+    end
+    return valid_players
+end
+
+--[[function mail.settings.mute_list.sync(name)
+    if minetest.get_modpath("beerchat") then
+        local players = {}
+        for other_player, _ in minetest.get_auth_handler().iterate() do
+            if beerchat.has_player_muted_player(name, other_player) then
+                table.insert(players, other_player)
+            end
+        end
+        return players
+    end
+    return nil
+end--]]
+
+function mail.settings.mute_list.transfer(name, value)
+    if minetest.get_modpath("beerchat") then
+        for other_player, _ in minetest.get_auth_handler().iterate() do -- unmute all
+            if not beerchat.execute_callbacks("before_mute", name, other_player) then
+                return false
+            end
+            minetest.get_player_by_name(name):get_meta():set_string(
+				"beerchat:muted:" .. other_player, "")
+        end
+        for _, other_player in ipairs(value) do -- then mute only players in table
+            minetest.get_player_by_name(name):get_meta():set_string(
+                    "beerchat:muted:" .. other_player, "true")
```

Comment:
This is bad idea as it will likely change to mod storage (because player meta is very limitated).
For this I'd recommend using API as it will then actually keep working after beerchat's internal storage is updated.

Reply:
Additionally it introduces another inconsistency as it depends on `/mute` command which is not definition of beerchat mute functionality.

Reply:
Okay because when I did it it was still the buggy muting API. I'm gonna make a commit in a second.

Reply:
EDIT : there's not yet a `mute_player` API command in `plugin/mute.lua`

Reply:
Muting API was not buggy for muting ever. Only for those shortcut API functions you can use to check if someone is muted.

Actual muting API uses event based checks, shortcut functions were not actually used anywhere (never called internally or externally) and were broken but those are only shortcuts for checking status. Muting is event based, not direct API call (because there's no single configuration for muting so you cannot store configuration for muting anywhere).

So, you implement check by adding this:
```lua
if minetest.get_modpath("beerchat") then
	beerchat.register_callback('before_send_pm', function(name, message, target)
		if mail.get_setting(target, "has_muted_"..name) then
			return false, "Optionally tell the player he can't send direct messages"
		end
	end)
end
```

Then use that check by calling (this shortcut was function buggy before but is now fixed):
this should be added where mail is actually sent
```lua
if beerchat.allow_private_message(from, to) then
	do_actually_deliver_mail_message_to_recipient(from, to, body)
end
```

Reply:
Also, you want that beerchat configrations keep affecting mail even after beerchat has been removed:
* This can be easily covered by utilizing `beerchat.allow_private_message(from, to)`. Or you can take a risk and keep syncing from player metadata but better only use this data AFTER beerchat is removed.

You want to sync muting information from mail mod towards beerchat:
* Better not do this but instead allow beechat to check if mail mod wants to mute someone instead of pushing your own stuff towards beerchat.
* If you want, beerchat `/list_muted` command can add functionality to list players muted by mail mod. Configuration sync between beerchat and mail mod is not needed for this. Mail mod does not have to implement its own API for this to work (code posted above is everything you need for this on mail mod side).
* If you want, mail mod can list players muted by any mod that utilized beerchat's muting API. Including server specific customizations. For this you can temporarily disable mail mods "before_send_pm" provider and call `beerchat.allow_private_message(from, to)` to get exact results from every mod connected to beerchat (several different ways to mute player).

I might be able to give other suggestions if there's specific functionality you'd want to have? If it is more just desire to sync because syncing sounds useful then we should probably find out what are the pros and cons for complete data sync.

Reply:
Ok for the suggestions (I already read them but didn't have the time to work on).

I can make a commit based on your suggestions, but I'm won't think it would be much better (we still have to get a copy for ourselves).

The idea is not to push to `master` a buggy PR, however my goal is to have #127 and #128 rebased to have a common base before merging new PRs (I don't want to treat a lot of merging conflicts).

Tell me if you're ok for the rebase. I already did a lot of tests, I fixed what I found, so for me it's stable actually.

Reply:
Comments are based on current PR target branch which currently is master.

Any rebases on development branches are always fine, no issues with that and should be done whenever you feel it helps with development.

But you can always switch that to another dev branch, also rebase on that dev branch if needed and it is actually what I do too when not sure about something, want to compare different somewhat complicated approaches or just want to get easy view to big picture about possible changes.

Reply:
Thanks for the answer.

For me, it is now operational (might be some lost bugs, as any software), but there's not so much issues at the moment.

It is a big pull request (gonna be rebased, with some squashes btw), so I would like other maintainers advice before doing it (such as @BuckarooBanzay )

Reply:
> For me, it is now operational (might be some lost bugs, as any software), but there's not so much issues at the moment.

:+1: 

> It is a big pull request (gonna be rebased, with some squashes btw), so I would like other maintainers advice before doing it (such as @BuckarooBanzay )

just squashing sounds fine, i would even do that into the `master` branch IMO (if the code is properly tested)